lib/wasi.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+/* global WebAssembly */
+const { Array, ArrayPrototype, Object } = primordials;
+const { ERR_INVALID_ARG_TYPE } = require('internal/errors').codes;
+const { emitExperimentalWarning } = require('internal/util');
+const { WASI: _WASI } = internalBinding('wasi');
+const kSetMemory = Symbol('setMemory');
+
+emitExperimentalWarning('WASI');
+
+
+class WASI {
+  constructor(options = {}) {
+    if (options === null || typeof options !== 'object')
+      throw new ERR_INVALID_ARG_TYPE('options', 'object', options);
+
+    // eslint-disable-next-line prefer-const
+    let { args, env, preopens } = options;
+
+    if (Array.isArray(args))
+      args = ArrayPrototype.map(args, (arg) => { return String(arg); });
+    else if (args === undefined)
+      args = [];
+    else
+      throw new ERR_INVALID_ARG_TYPE('options.args', 'Array', args);
+
+    const envPairs = [];
+
+    if (env !== null && typeof env === 'object') {
+      for (const key in env) {
+        const value = env[key];
+        if (value !== undefined)
+          envPairs.push(`${key}=${value}`);
+      }
+    } else if (env !== undefined) {
+      throw new ERR_INVALID_ARG_TYPE('options.env', 'Object', env);
+    }
+
+    const preopenArray = [];
+
+    if (typeof preopens === 'object' && preopens !== null) {
+      Object.keys(preopens).forEach((key) => {
+        preopenArray.push(String(key));
+        preopenArray.push(String(preopens[key]));
+      });
+    } else if (preopens !== undefined) {
+      throw new ERR_INVALID_ARG_TYPE('options.preopens', 'Object', preopens);
+    }
+
+    const wrap = new _WASI(args, envPairs, preopenArray);
+
+    for (const prop in wrap) {
+      wrap[prop] = wrap[prop].bind(wrap);
+    }
+
+    this[kSetMemory] = wrap._setMemory;
+    delete wrap._setMemory;
```

Comment:
Afaict this line is a no-op because `_setMemory` is on the prototype, not the instance?

If you want to keep this more private, you can make `_setMemory` a function that takes a `_WASI` argument instead of using it as the `args.This()` value, similar to e.g. `MessagePort::Drain()` in `node_messaging.cc`

Reply:
It wasn't quite a no-op. It was deleting the bound function, which is what I wanted to hide. Moving `_setMemory()` to an instance method appears to make this work exactly how I originally wanted now.